Allow serving API through TLS w/ client authentication

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1503 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 3e2aeb8f2158733a1825cccec58a029cf2a411f6
  :round_pushpin: **create new module for generate TLS settings for client auth**
  This is mostly copied from cardano-sl and re-arranged a bit. This validation is very
basic, but so is our TLS setup anyway. In a new module because it's quite isolated
while quite verbose.

- dfcf26642e28861db25cf6352c7901f39a44d897
  :round_pushpin: **run server socket through TLS when provided with a TLS configuration**
  Passing 'Nothing' as an argument will preserve the current behavior where the socket run with
no TLS activated. This is still useful for current integration tests and for the Jörmungandr
setup.

- f0afdcf8419e6b9e2f57feac45282f8f58782e03
  :round_pushpin: **add command-line option for parsing TLS configuration**
  Very simple here, we simply go with path and _assume_ that the given PKI is somewhat
well-formed. It's hard to do any sort of validation here but it'll be good to investigate
what sort of runtime errors we might run into if the provided files aren't certificates...

- 29182d5543cfad19c42f2f47dd120ce77e7da16a
  :round_pushpin: **unit test TLS settings with a mock server**
  Certificates generated using cardano-sl-x509. The test setup is fairly simple, a webserver
is started using the very same 'start' method we use for starting the wallet server.
Then, we simply try to ping it in various scenarios, with and without TLS enabled on the
client side.

- c1d8ea17222108eb54dfe0f9053d1569995f865f
  :round_pushpin: **re-generate nix machinery**

# Comments

<!-- Additional comments or screenshots to attach if any -->

:warning: Haven't tested any of this _yet_ ! I'll see if we can simply enable this on our integration setup. 

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
create new module for generate TLS settings for client auth

This is mostly copied from cardano-sl and re-arranged a bit. This validation is very
basic, but so is our TLS setup anyway. In a new module because it's quite isolated
while quite verbose.

KtorZ committed:
run server socket through TLS when provided with a TLS configuration

Passing 'Nothing' as an argument will preserve the current behavior where the socket run with
no TLS activated. This is still useful for current integration tests and for the Jörmungandr
setup.

KtorZ committed:
add command-line option for parsing TLS configuration

Very simple here, we simply go with path and _assume_ that the given PKI is somewhat
well-formed. It's hard to do any sort of validation here but it'll be good to investigate
what sort of runtime errors we might run into if the provided files aren't certificates...

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server/Tls.hs:
```diff
@@ -0,0 +1,99 @@
+    -- This checks makes sense only for remote services, to validate that the
+    -- fully qualified hostname from the certificate matches the one from the
+    -- service we're trying to reach. This is of little use for a server
+    -- validation.
```
:+1:

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server/Tls.hs:
```diff
@@ -0,0 +1,99 @@
+
+    -- By default, X509.Validation validates the certificate names against the host
+    -- which is irrelevant when checking the client certificate (but relevant for
+    -- the client when checking the server's certificate).
```
:+1:

paweljakubas reviewed (approved):
looks reasonable, here assumptions why to omit some validations. PR needs some nix patch application. But would be great to add **some testing** for it to gain confidence it works as expected

KtorZ committed:
re-generate nix machinery

rhyslbw commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1672,6 +1673,28 @@ addressIndexOption = optionT $ mempty
+    <*> tlsSvKeyOption
+  where
+    tlsCaCertOption = optionT $ mempty
+        <> long "--tls-ca-cert"
```
```suggestion
        <> long "tls-ca-cert"
```

KtorZ replied:
Whola :man_facepalming:

KtorZ replied:
commit [amended](https://github.com/input-output-hk/cardano-wallet/compare/c1d8ea17222108eb54dfe0f9053d1569995f865f..c6cad2975fa580ff662ea0a440c441e3c5c8a3df)

rhyslbw commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1672,6 +1673,28 @@ addressIndexOption = optionT $ mempty
+        <> help "A x.509 Certificate Authority (CA) certificate."
+
+    tlsSvCertOption = optionT $ mempty
+        <> long "--tls-sv-cert"
```
```suggestion
        <> long "tls-sv-cert"
```

KtorZ replied:
commit [amended](https://github.com/input-output-hk/cardano-wallet/compare/c1d8ea17222108eb54dfe0f9053d1569995f865f..c6cad2975fa580ff662ea0a440c441e3c5c8a3df)

rhyslbw commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1672,6 +1673,28 @@ addressIndexOption = optionT $ mempty
+        <> help "A x.509 Server (SV) certificate."
+
+    tlsSvKeyOption = optionT $ mempty
+        <> long "--tls-sv-key"
```
```suggestion
        <> long "tls-sv-key"
```

KtorZ replied:
commit [amended](https://github.com/input-output-hk/cardano-wallet/compare/c1d8ea17222108eb54dfe0f9053d1569995f865f..c6cad2975fa580ff662ea0a440c441e3c5c8a3df)

rhyslbw reviewed (dismissed):
Currently
```
Available options:
  -h,--help                Show this help text
  --help-tracing           Show help for tracing options
  --listen-address HOST    Specification of which host to the bind API server
                           to. Can be an IPv[46] address, hostname, or
                           '*'. (default: 127.0.0.1)
  --random-port            serve wallet API on any available port (conflicts
                           with --port)
  --port INT               port used for serving the wallet API. (default: 8090)
  ----tls-ca-cert FILE     A x.509 Certificate Authority (CA) certificate.
  ----tls-sv-cert FILE     A x.509 Server (SV) certificate.
  ----tls-sv-key FILE      The RSA Server key which signed the x.509 server
                           certificate.
...
```

KtorZ committed:
unit test TLS settings with a mock server

Certificates generated using cardano-sl-x509. The test setup is fairly simple, a webserver
is started using the very same 'start' method we use for starting the wallet server.
Then, we simply try to ping it in various scenarios, with and without TLS enabled on the
client side.
